[nnc] Test ability to vectorize reads from an intermediate tensor

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53686 [nnc] Add an initialization expression to Reduce()
* **#53552 [nnc] Test ability to vectorize reads from an intermediate tensor**

This test doesn't work today because we don't properly vectorize
"FunctionCall" (which is the way one accesses an intermediate tensor).

Differential Revision: [D26895550](https://our.internmc.facebook.com/intern/diff/D26895550/)

bertmaher committed:
[nnc] Test ability to vectorize reads from an intermediate tensor

This test doesn't work today because we don't properly vectorize
"FunctionCall" (which is the way one accesses an intermediate tensor).

Differential Revision: [D26895550](https://our.internmc.facebook.com/intern/diff/D26895550/)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Test ability to vectorize reads from an intermediate tensor"


This test doesn't work today because we don't properly vectorize
"FunctionCall" (which is the way one accesses an intermediate tensor).

Differential Revision: [D26895550](https://our.internmc.facebook.com/intern/diff/D26895550/)

[ghstack-poisoned]

navahgar commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -3921,5 +3921,29 @@ TEST(LoopNest, DISABLED_ColReduceSplitMaskUnevenReorder) {
+  auto loops = nest.getLoopStmtsFor(b);
+  nest.vectorize(loops[0]);
+  loops = nest.getLoopStmtsFor(c);
+  nest.vectorize(loops[0]);
```
I don't think this case ever happens in the regular flow in NNC because we "inline" all such calls before vectorization. See the call here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/kernel.cpp#L1567

The argument then is - are we doing the right thing? So far, we have been inlining everything, so this would never happen. But may be needed going forward.

Are you seeing real use of such flows (where we don't inlining the intermediate tensors)?

bertmaher replied:
Yeah the real use case here is to support convolution with padding.  This isn't the only way to do it but essentially I want to:
```
auto paddedInput = Compute(...);
auto conv = Compute(/* use paddedInput */);
// Apply some scheduling primitives to conv
```

bertmaher committed:
Update on "[nnc] Test ability to vectorize reads from an intermediate tensor"


This test doesn't work today because we don't properly vectorize
"FunctionCall" (which is the way one accesses an intermediate tensor).

Differential Revision: [D26895550](https://our.internmc.facebook.com/intern/diff/D26895550/)

[ghstack-poisoned]

ZolotukhinM commented:
Once #54998 lands, the test should work (with a minor change in how you construct the loopnest):
```
-  LoopNest nest({c});
+  LoopNest nest({c}, {b, c});
```
